File: 프로그래머스/2/87946. 피로도/피로도.py
from itertools import permutations

def solution(k, dungeons):
    answer = 0
    
    for game in permutations(dungeons, len(dungeons)):
        game_k = k
        count = 0
        
        for need, used in game:
            if need <= game_k:
                game_k -= used
                count += 1
        
        if answer < count: answer = count
    
    return answer

File: 프로그래머스/2/87946. 피로도/피로도.js
function solution(k, dungeons) {
    let answer = 0;
    const dungeons_dfs = [];
    
    const dfs = (arr, visited=[]) => {
        if (visited.length === dungeons.length) {
            dungeons_dfs.push([...visited]);
        }
        else {
            for (let i = 0; i < arr.length; i++) {
                visited.push(arr[i]);
                dfs(arr.filter((_, index) => index !== i), visited);
                visited.pop();
            }
        }
    };
    
    dfs(dungeons);
    
    dungeons_dfs.forEach(game => {
        let game_k = k;
        let count = 0;
        
        game.forEach(([need, used]) => {
            if (need <= game_k) {
                game_k -= used;
                count++;
            }
        });
        
        if (answer < count) answer = count;
    });
    
    return answer;
}